main.py

```diff
@@ -0,0 +1,449 @@
+        depth_img = depth_img * robot.cam_depth_scale # Apply depth scale from calibration
+
+        # Get heightmap from RGB-D image (by re-projecting 3D point cloud)
+        color_heightmap, depth_heightmap = utils.get_heightmap(color_img, depth_img, robot.cam_intrinsics, robot.cam_pose, workspace_limits, heightmap_resolution)
+        valid_depth_heightmap = depth_heightmap.copy()
+        valid_depth_heightmap[np.isnan(valid_depth_heightmap)] = 0
+
+        # Save RGB-D images and RGB-D heightmaps
+        logger.save_images(trainer.iteration, color_img, depth_img, '0')
+        logger.save_heightmaps(trainer.iteration, color_heightmap, valid_depth_heightmap, '0')
+
+        # Reset simulation or pause real-world training if table is empty
+        stuff_count = np.zeros(valid_depth_heightmap.shape)
+        stuff_count[valid_depth_heightmap > 0.02] = 1
+        empty_threshold = 300
+        if is_sim and is_testing:
+            empty_threshold = 10  
+        if np.sum(stuff_count) < empty_threshold or (is_sim and no_change_count[0] + no_change_count[1] > 10):
+            no_change_count = [0, 0]
+            if is_sim:
+                print('Not enough objects in view (value: %d)! Repositioning objects.' % (np.sum(stuff_count)))
+                robot.restart_sim()
+                robot.add_objects()
+                if is_testing: # If at end of test run, re-load original weights (before test run)
+                    trainer.model.load_state_dict(torch.load(snapshot_file))
+            else:
+                # print('Not enough stuff on the table (value: %d)! Pausing for 30 seconds.' % (np.sum(stuff_count)))
+                # time.sleep(30)
+                print('Not enough stuff on the table (value: %d)! Flipping over bin of objects...' % (np.sum(stuff_count)))
+                robot.restart_real()
+
+            trainer.clearance_log.append([trainer.iteration]) 
+            logger.write_to_log('clearance', trainer.clearance_log)
+            if is_testing and len(trainer.clearance_log) >= max_test_trials:
+                exit_called = True # Exit after training thread (backprop and saving labels)
+            continue
+
+        if not exit_called: 
+
+            # Run forward pass with network to get affordances
+            push_predictions, grasp_predictions, state_feat = trainer.forward(color_heightmap, valid_depth_heightmap, is_volatile=True)
+
+            # Execute best primitive action on robot in another thread
+            nonlocal_variables['executing_action'] = True
+
+        # Run training iteration in current thread (aka training thread)
+        if 'prev_color_img' in locals():
+
+            # Detect changes
+            depth_diff = abs(depth_heightmap - prev_depth_heightmap)
+            depth_diff[np.isnan(depth_diff)] = 0
+            depth_diff[depth_diff > 0.3] = 0
+            depth_diff[depth_diff < 0.01] = 0
+            depth_diff[depth_diff > 0] = 1
+            change_threshold = 300
```

Comment:
How did they arrive at the 300 value for change_threshold? What does this value mean?

Reply:
Trial and error.